From 0dfc8e2c8e1455ae872958f80a7a81b515ec6426 Mon Sep 17 00:00:00 2001
From: Mao Minkai <maominkai@wxiat.com>
Date: Tue, 7 Jun 2022 17:41:34 +0800
Subject: [PATCH] sw64: mm: reorder memblock_init process
Patch-mainline: NO, OTHERS
References: openEuler-22.03-LTS
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I56P72

--------------------------------

Reorder memblock_init process to make sure kernel text range and initrd
range won't be removed by cmdline.

Signed-off-by: Mao Minkai <maominkai@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/mm/init.c | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/arch/sw_64/mm/init.c b/arch/sw_64/mm/init.c
index d0e934356dd5..e3c2a20e54a2 100644
--- a/arch/sw_64/mm/init.c
+++ b/arch/sw_64/mm/init.c
@@ -193,15 +193,27 @@ void __init sw64_memblock_init(void)
 
 	memblock_remove(1ULL << MAX_PHYSMEM_BITS, PHYS_ADDR_MAX);
 
-	/* Make sure kernel text is in memory range. */
-	memblock_add(__pa_symbol(_text), (unsigned long)(_end - _text));
-	memblock_reserve(__pa_symbol(_text), _end - _text);
-
 	max_pfn = max_low_pfn = PFN_DOWN(memblock_end_of_DRAM());
 
 	memblock_allow_resize();
 	memblock_initialized = true;
 	process_memmap();
+
+	/* Make sure kernel text is in memory range. */
+	memblock_add(__pa_symbol(_text), _end - _text);
+	memblock_reserve(__pa_symbol(_text), _end - _text);
+
+	/* Make sure initrd is in memory range. */
+	if (sunway_boot_params->initrd_start) {
+		phys_addr_t base = __pa(sunway_boot_params->initrd_start);
+		phys_addr_t size = sunway_boot_params->initrd_size;
+
+		memblock_add(base, size);
+		memblock_reserve(base, size);
+	}
+
+	/* end of DRAM range may have been changed */
+	max_pfn = max_low_pfn = PFN_DOWN(memblock_end_of_DRAM());
 }
 
 #ifndef CONFIG_NUMA
-- 
2.34.1

